### ISSUE TITLE
hasDirtyAttributes: true right after findAll

### ISSUE THREAD

**Issue by house-md** [ghidoz/angular2-jsonapi#167]
Hello I have problem with loading data, right after I load them from backend, entity has **hasDirtyAttributes:true** and on save its sending complete entity back to backend instead of only changed fields. happens for **findAll()** and **find()**

![screenshot_7](https://user-images.githubusercontent.com/7100025/38515905-f5a3dc56-3c35-11e8-9293-52df6663c726.png)
**list.component.ts**
```js
ngOnInit() {
this.clients = this.route.snapshot.data['clients'].getModels();
}
```

**client.resolve.ts**
```js
export class ClientsResolve implements Resolve<any> {
  constructor(private datastore: ClientDatastore) {
  }

  resolve(route: ActivatedRouteSnapshot): Observable<any> | Promise<any> | any {

    let include = '';
    if (route.data.include) {
      include += route.data.include;
    }
    return this.datastore.findAll(Client, {
      include: include
    });
  }
}
```

**client.routes.ts**
```js
{
    path: 'list',
    component: ClientListComponent,
    data: {
      breadcrumb: 'Přehled',
      include: 'advisor'
    },
    resolve: {
      clients: ClientsResolve
    }
  },
```

**client.datastore.ts**
```js
const config: DatastoreConfig = {
  baseUrl: environment.apiURL + '/v2',
  models: {
    client: Client,
    advisors: Advisor,
    goals: Goal
  }
};

@Injectable()
@JsonApiDatastoreConfig(config)
export class ClientDatastore extends JsonApiDatastore {
  constructor(http: Http) {
    super(http);
    this.headers = new Headers({'Authorization': 'Bearer ' + localStorage.getItem('token')});
  }
}
```

**Comment by house-md** [comment]
@abrararshad after some research I found that this problem is caused by this commit where is absolutely ignored if its new instance created from query or not
https://github.com/ghidoz/angular2-jsonapi/commit/bb1bfdf2c40ee26f530839c2e227519b91019611

**Comment by Birdr518** [comment]
I can confirm this issue. Using Version 4.1.0 all attributes of my data objects are considered as "dirtyAttributes" - even right after loading them from the backend. 

 

```

 this.dataStore.findRecord(ApiUsers, "", {}, this.authenticationHeader, 
      ApiStringsService.baseUri + "/my/profile"
    ).subscribe((apiUser: ApiUsers) => {

      console.log(apiUser.hasDirtyAttributes);  // -> true
      apiUser.email = "bla";
      console.log(apiUser.hasDirtyAttributes);` // -> true

});
```


Using Version 4.0.0 this issue is gone and the first `console.log` will output `false`.
 
But still, if i have an attribute in the form like this:

```
 @Attribute()
  person:
    {
      title: string,
      firstname: string,
      lastname: string,
      fullname: string,
      gender: number,
      birthday: {
        date: Date,
        timezone_type: number,
        timezone: string
      }
    };
```


changing one nested attribute like `person.title` not only `person.title` isn't considered as a "dirty attribute", nor its parent `person` (the whole attribute) is stated as "dirty attribute".
Is this behaviour intended?

**Comment by house-md** [comment]
@Birdr518 I am currently working on fix and also feature to work better with therse nested attributes. I hope i will be able to create pull request in next week.

**Comment by house-md** [comment]
Hello @RicardoNeves and @MarkCorneth.
I am sorry for my late reply. In branch  #167-hasDirtyAttributes is my actual implementation. There is a little bit reworked dirty checking and working with old/newValue.

I also added nested array/object attributes. I am currently testing this solution for nested attributes. I am not happy with it but after maybe 10 different aproaches but this one seems working.

I tried mupltiple solutions with Proxy array/objects but in the end it always had some kind of bug or ended up with infinite loops of getters and apply methods (I still have code to this solution so I will try to finish it but for now this one should be also working. There is one setback -> for nested attributes I had to put their dirtyChecking to method in model -> hasDirtyAttributes which is always called before save and can be triggered by user any time.
